core/amber/src/main/python/core/models/RTupleExecutor.py

```diff
@@ -0,0 +1,193 @@
+
+        :param tuple_: Tuple, a Tuple from an input port to be processed.
+        :param port: int, input port index of the current Tuple.
+        :return: Iterator[Optional[TupleLike]], producing one TupleLike object at a
+            time, or None.
+        """
+        with local_converter(arrow_converter):
+            input_schema: pa.Schema = tuple_._schema.as_arrow_schema()
+            input_fields: list[str] = [field.name for field in input_schema]
+            non_binary_fields: list[str] = [
+                field.name for field in input_schema if field.type != pa.binary()
+            ]
+            binary_fields: list[str] = [
+                field.name for field in input_schema if field.type == pa.binary()
+            ]
+
+            non_binary_tuple: Tuple = tuple_.get_partial_tuple(non_binary_fields)
+            non_binary_tuple_schema: pa.Schema = (
+                non_binary_tuple._schema.as_arrow_schema()
+            )
+            non_binary_pyarrow_array: pa.StructArray = pa.array(
+                [non_binary_tuple.as_dict()], type=pa.struct(non_binary_tuple_schema)
+            )
+
+            binary_tuple: Tuple = tuple_.get_partial_tuple(binary_fields)
+            binary_r_list: dict[str, object] = {}
+            for k, v in binary_tuple.as_dict().items():
+                if isinstance(v, bytes):
+                    binary_r_list[k] = pickle.loads(v[10:])
+                elif isinstance(v, datetime.datetime):
+                    binary_r_list[k] = robjects.vectors.POSIXct.sexp_from_datetime([v])
+                else:
+                    binary_r_list[k] = v
+
+            binary_r_list: rpy2.robjects.ListVector = robjects.vectors.ListVector(
+                binary_r_list
+            )
+
+            input_r_list: rpy2.robjects.ListVector = (
+                RTupleExecutor._combine_binary_and_non_binary_lists(
+                    non_binary_pyarrow_array, binary_r_list
+                )
+            )
+
+            output_r_generator: rpy2.robjects.SignatureTranslatedFunction = self._func(
+                input_r_list, port
+            )
+
+            while True:
+                output_r_tuple: rpy2.robjects.ListVector = output_r_generator()
+                if (
+                    isinstance(output_r_tuple, rinterface.SexpSymbol)
+                    and str(output_r_tuple) == ".__exhausted__."
+                ):
+                    break
+                if isinstance(output_r_tuple.names, rpy2.rinterface_lib.sexp.NULLType):
+                    yield None
+                    break
+
+                diff_fields: list[str] = [
+                    field_name
+                    for field_name in output_r_tuple.names
+                    if field_name not in input_fields
+                ]
+
+                output_python_dict: dict[str, object] = {
+                    key: output_r_tuple.rx2(key) for key in (input_fields + diff_fields)
+                }
```

Comment:
the order of the fields could be dropped. can we preserve it as well?

Reply:
Yes, I forgot to mention this to you. Before I start this operation I have a list called "input_fields" which creates a list of fields of the input tuple, preserving the original order. Then, I create a new variable called "diff_fields" which is a list of new fields that the user adds (if any). diff_fields is then appended to the input_fields list and is then iterated over to create the new output Tuple.

Reply:
Just to confirm, r tuple's fields order is always the same as the python tuple's fields order right?

Reply:
Yes, that is correct